### DISCUSSION TITLE
Allow fetch_links depth to be specified

### DISCUSSION THREAD

**Discussion by rosenfeld-ecore** [BeanieODM/beanie#690] [Feature Request]
Hello, I'm working on some models with a deep level of links all around and fetching them all means a lot of data to fetch, making the full query/fetch quite slow (~600ms).

I tried using projections to select the fields I'm interested in, however, since some fields are links, and those have also links to other models, it doesn't help much in those cases. It would be great if we could provide the depth level in the projection, for example, to specify how many levels should be fetched for the particular link in the projection.

Even outside of projections, it would be great to be able to specify the depth length when fetching links.

Would it be possible to add a feature like that to beanie?

**Comment by mrrobinhood5** [comment]
I concur with this feature. My current workaround is to not fetch links or make an async generator to fetch one result at a time.

**Comment by akloster** [comment]
I would caution that having that many relations may not be a good way of using Mongodb. Embedding is often preferrable to relations, and if you have some experience with SQL data modeling, you'll likely overuse relations.

One issue with resolving many levels of relations is that it requires multiple pipeline stages. First problem with limiting those levels is that you'd need some way to specify how many levels. But also, would you specify to go down all relations to the same degree, or just a subset?
Entity Framework (Dotnet) solves this by ".include" and ".thenInclude" methods on their queries. So you only get what you need. However, this often results in things being null or empty liste when you least expect it.

I think for the time being the only workaround is to find better custom queries. When you ask Mongodb/Beanie to assemble a complicated structure of nested relations, you may be performing too much logic on the Python side of things. Try to perform the logic, both queries and updates, more inside the database. You can use custom pipelines to project to your pydantic models, and use update queries to only update the parts of the data that need to change.

**Comment by rosenfeld-ecore** [comment]
@akloster the main issue for me is that such queries would be very common with our data models. Embedding wouldn't work well for many reasons in our domain model and we're often only interested in one level deep, usually containing a list of links. For example, a space would contain a list of sensors and we'd like to get the space sensor ids and names besides the space information, but the space is also linked to other models which are not relevant to this query. Even if we use projections, as far as I understand beanie's code, I think beanie will build the lookup pipeline to include all links even though we're just projecting the sensors information. I often miss some way to specify which links I'd like beanie to include in the aggregate pipeline, so that it doesn't aggregate so many data when we don't need it.

**Reply by rosenfeld-ecore** [reply]
Also, I tried to add the `Settings` class to include the projections to the information I'm interested in, but I noticed it won't traverse linked documents unless I pass `fetchLinks=True`. Another option would be for beanie to accept an array for the `fetchLinks` param, such as `fetchLinks=[SpacesModel.sensors]`, for example.
